lib/auth/machineid/machineidv1/spiffe_federation_syncer.go

```diff
@@ -0,0 +1,631 @@
+	// configured upstream trust domain from causing excessive staleness in the
+	// local cluster.
+	MaxSyncInterval time.Duration
+	// DefaultSyncInterval is the interval between syncs that will be used if an
+	// upstream trust domain does not specify a refresh hint.
+	DefaultSyncInterval time.Duration
+
+	// SyncTimeout is the maximum time that a sync operation is allowed to take.
+	// If a sync operation takes longer than this value, it will be aborted and
+	// retried.
+	SyncTimeout time.Duration
+
+	// SPIFFEFetchOptions are the options that will be used when fetching a
+	// trust bundle from a remote cluster. These options will be passed to the
+	// spiffebundle.FetchBundle function. This is usually used during testing
+	// to override the Web PKI CAs.
+	SPIFFEFetchOptions []federation.FetchOption
+}
+
+// SPIFFEFederationSyncer is a syncer that manages the trust bundles of
+// federated clusters. It runs on a single elected auth server.
+type SPIFFEFederationSyncer struct {
+	cfg SPIFFEFederationSyncerConfig
+}
+
+const (
+	minRefreshInterval = time.Minute * 1
+	maxRefreshInterval = time.Hour * 24
+	// SPIFFE Federation (4.1):
+	// > If not set, a reasonably low default value should apply - five minutes
+	// > is recommended
+	defaultRefreshInterval = time.Minute * 5
+	defaultSyncTimeout     = time.Second * 30
+)
+
+// NewSPIFFEFederationSyncer creates a new SPIFFEFederationSyncer.
+func NewSPIFFEFederationSyncer(cfg SPIFFEFederationSyncerConfig) (*SPIFFEFederationSyncer, error) {
+	switch {
+	case cfg.Backend == nil:
+		return nil, trace.BadParameter("backend: must be non-nil")
+	case cfg.Store == nil:
+		return nil, trace.BadParameter("store: must be non-nil")
+	case cfg.Logger == nil:
+		return nil, trace.BadParameter("logger: must be non-nil")
+	case cfg.Clock == nil:
+		return nil, trace.BadParameter("clock: must be non-nil")
+	}
+	if cfg.MinSyncInterval == 0 {
+		cfg.MinSyncInterval = minRefreshInterval
+	}
+	if cfg.MaxSyncInterval == 0 {
+		cfg.MaxSyncInterval = maxRefreshInterval
+	}
+	if cfg.DefaultSyncInterval == 0 {
+		cfg.DefaultSyncInterval = defaultRefreshInterval
+	}
+	if cfg.SyncTimeout == 0 {
+		cfg.SyncTimeout = defaultSyncTimeout
+	}
+	return &SPIFFEFederationSyncer{
+		cfg: cfg,
+	}, nil
+}
+
+func (s *SPIFFEFederationSyncer) Run(ctx context.Context) error {
+	// Loop to retry if acquiring lock fails, with some backoff to avoid pinning
+	// the CPU.
+	waitWithJitter := retryutils.NewSeventhJitter()(time.Second * 10)
+	for {
+		err := backend.RunWhileLocked(ctx, backend.RunWhileLockedConfig{
+			LockConfiguration: backend.LockConfiguration{
+				Backend:  s.cfg.Backend,
+				LockName: "spiffe_federation_syncer",
+				TTL:      time.Minute,
+				// It doesn't matter too much if the syncer isn't running for
+				// a short period of time so we can take a relaxed approach to
+				// retrying to grab the lock.
+				RetryInterval: time.Second * 30,
+			},
+		}, s.syncTrustDomains)
+		if ctx.Err() != nil {
+			return nil
+		}
```

Comment:
It was mostly to avoid sending a spurious error log about the context cancellation but I can see how it just makes this bit of code unnecessary. I'll drop it.

Reply:
49b46d984f